src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+from functools import partial
+from typing import Any, Callable, NamedTuple, Optional, Union
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+from jax import grad, jit, lax, random
+from jaxopt import OptStep
+from jaxopt._src import loop
+from jaxopt.prox import prox_none
+
+from .tree_utils import tree_add_scalar_mul, tree_l2_norm, tree_slice, tree_sub
+from .typing import KeyArrayLike, Pytree
+
+
+class SVRGState(NamedTuple):
+    iter_num: int
+    key: KeyArrayLike
+    error: float
+    stepsize: float
+    xs: Optional[Pytree] = None
+    df_xs: Optional[Pytree] = None
+
+
+class ProxSVRG:
+    """
+    Prox-SVRG solver
+
+    Borrowing from jaxopt.ProximalGradient, this solver minimizes:
+
+      objective(params, hyperparams_prox, *args, **kwargs) =
+        fun(params, *args, **kwargs) + non_smooth(params, hyperparams_prox)
+
+    Attributes
+    ----------
+    fun: Callable
+        Smooth function of the form ``fun(x, *args, **kwargs)``.
+    prox: Callable
+        Proximal operator associated with the function ``non_smooth``.
+        It should be of the form ``prox(params, hyperparams_prox, scale=1.0)``.
+        See ``jaxopt.prox`` for examples.
+    maxiter : int
+        Maximum number of epochs to run the optimization for.
+    key : jax.random.PRNGkey
+        jax PRNGKey to start with. Used for sampling random data points.
+    stepsize : float
+        Constant step size to use.
+    tol: float
+        Tolerance level for the error when comparing parameters
+        at the end of consecutive epochs to check for convergence.
+    batch_size: int
+        Number of data points to sample per inner loop iteration.
+
+    Example
+    -------
+    def loss_fn(params, X, y):
+        ...
+
+    svrg = ProxSVRG(loss_fn, prox_fun)
+    params, state = svrg.run(init_params, prox_lambda, X, y)
+
```

Comment:
Example is not rendered correctly. I believe it should be [formatted like](https://numpydoc.readthedocs.io/en/latest/format.html#examples):
```suggestion
    Examples
    --------
    >>> def loss_fn(params, X, y):
    >>>    ...
    >>>
    >>> svrg = ProxSVRG(loss_fn, prox_fun)
    >>> params, state = svrg.run(init_params, prox_lambda, X, y)

```

Reply:
Changed it.